neqo-common/src/codec.rs

```diff
@@ -92,33 +82,43 @@ impl<'a> Decoder<'a> {
         Some(res)
     }
 
-    /// Decodes an unsigned integer of length 1..=8.
-    ///
-    /// # Panics
-    ///
-    /// This panics if `n` is not in the range `1..=8`.
-    pub fn decode_uint(&mut self, n: usize) -> Option<u64> {
-        assert!(n > 0 && n <= 8);
+    #[inline]
+    pub(crate) fn decode_n(&mut self, n: usize) -> Option<u64> {
```

Comment:
Given that `n` is known at compile time, i.e. derived from `size_of::<T>()`, one could leverage const generics here:

```suggestion
    pub(crate) fn decode_n<const N: usize>(&mut self) -> Option<u64> {
```

That said, I don't think it has a performance gain:

```
➜  critcmp baseline const                                                                                                  
group                            baseline                                const
-----                            --------                                -----
decode 1024 bytes, mask 3f       1.00  1066.4±116.06ns        ? ?/sec    1.02   1092.5±2.61ns        ? ?/sec
decode 1024 bytes, mask 7f       1.00      3.5±0.15µs        ? ?/sec     1.01      3.5±0.00µs        ? ?/sec
decode 1024 bytes, mask ff       1.00      2.3±0.00µs        ? ?/sec     1.00      2.3±0.00µs        ? ?/sec
decode 1048576 bytes, mask 3f    1.00   1112.8±2.42µs        ? ?/sec     1.00  1116.0±11.03µs        ? ?/sec
decode 1048576 bytes, mask 7f    1.00      3.3±0.40ms        ? ?/sec     1.10      3.6±0.01ms        ? ?/sec
decode 1048576 bytes, mask ff    1.00      2.3±0.01ms        ? ?/sec     1.00      2.3±0.00ms        ? ?/sec
decode 32768 bytes, mask 3f      1.00     34.8±0.20µs        ? ?/sec     1.00     34.8±0.44µs        ? ?/sec
decode 32768 bytes, mask 7f      1.00    112.1±0.18µs        ? ?/sec     1.00    112.0±0.14µs        ? ?/sec
decode 32768 bytes, mask ff      1.00     73.7±0.12µs        ? ?/sec     1.00     73.8±0.11µs        ? ?/sec
```

Reply:
Just the generic won't lead to improvements (the compiler should be smart enough), but I think optimized code per length might help.

Reply:
Yeah, I tried the const generic :)  The n==1 optimization below is part of that.